File: x-runtime/python/sdk/src/hermes_worker/transport.py
"""Zero-copy local transport via memory-mapped ring buffer."""
from __future__ import annotations

import mmap
import os
import struct
import threading
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable, Iterator, Optional


RING_MAGIC = b"HMRING00"
RING_VERSION = 1
HEADER_SIZE = 128
MIN_CAPACITY = 4096
ALIGN = 64


@dataclass
class TransportCapabilities:
    supports_zero_copy: bool = True
    max_message_bytes: int = 4096
    ring_pages: int = 1


class ZeroCopyRing:
    def __init__(self, path: Path | str, capacity: int = MIN_CAPACITY) -> None:
        self.path = Path(path)
        self.capacity = max(MIN_CAPACITY, capacity)
        self._lock = threading.Lock()
        self._mm: Optional[mmap.mmap] = None
        self._create()

    def _create(self) -> None:
        fd = os.open(str(self.path), os.O_RDWR | os.O_CREAT | os.O_TRUNC)
        try:
            ring_size = HEADER_SIZE + self.capacity
            os.ftruncate(fd, ring_size)
            self._mm = mmap.mmap(fd, ring_size)
            self._write_header(ring_size)
        finally:
            os.close(fd)

    def _write_header(self, ring_size: int) -> None:
        if self._mm is None:
            raise RuntimeError("mmap not initialized")
        self._mm.seek(0)
        self._mm.write(RING_MAGIC)
        self._mm.write(struct.pack("<I", RING_VERSION))
        self._mm.write(struct.pack("<I", 0))
        self._mm.write(struct.pack("<Q", ring_size))
        self._mm.write(struct.pack("<I", self.capacity))
        self._mm.write(struct.pack("<I", 0))
        self._mm.write(struct.pack("<I", 0))
        self._mm.write(struct.pack("<I", 0))
        self._mm.write(struct.pack("<Q", 0))
        self._mm.write(struct.pack("<I", 1))
        self._mm.write(b"\x00" * (HEADER_SIZE - 40))

    @classmethod
    def open(cls, path: Path | str) -> ZeroCopyRing:
        path = Path(path)
        fd = os.open(str(path), os.O_RDWR)
        try:
            size = os.fstat(fd).st_size
            if size < HEADER_SIZE + MIN_CAPACITY:
                raise RuntimeError("ring too small")
            mm = mmap.mmap(fd, size)
            magic = mm.read(8)
            version = struct.unpack("<I", mm.read(4))[0]
            if magic != RING_MAGIC or version != RING_VERSION:
                raise RuntimeError("invalid ring magic/version")
            mm.seek(24)
            capacity = struct.unpack("<I", mm.read(4))[0]
            ring = object.__new__(cls)
            ring.path = path
            ring.capacity = capacity
            ring._lock = threading.Lock()
            ring._mm = mm
            return ring
        finally:
            os.close(fd)

    def send(self, payload: bytes) -> None:
        if len(payload) > self.capacity - 4:
            raise ValueError("payload too large")
        with self._lock:
            if self._mm is None:
                raise RuntimeError("mmap not initialized")
            head = struct.unpack("<I", self._mm[32:36])[0]
            tail = struct.unpack("<I", self._mm[36:40])[0]
            closed = struct.unpack("<I", self._mm[40:44])[0]
            if closed:
                raise RuntimeError("ring closed")
            needed = 4 + len(payload)
            free = (self.capacity - 1 - ((head - tail) % self.capacity)) % self.capacity
            if free < needed:
                raise RuntimeError("ring full")
            offset = HEADER_SIZE + (head % self.capacity)
            if offset + needed > HEADER_SIZE + self.capacity:
                chunk = HEADER_SIZE + self.capacity - offset
                self._mm[offset:offset + chunk] = payload[:chunk] + struct.pack("<I", len(payload))
                self._mm[HEADER_SIZE:HEADER_SIZE + needed - chunk] = b"\x00" * (needed - chunk - 4) + payload[chunk - needed:]
            else:
                self._mm[offset:offset + 4] = struct.pack("<I", len(payload))
                self._mm[offset + 4:offset + needed] = payload
            self._mm[32:36] = struct.pack("<I", (head + needed) % self.capacity)

    def receive(self) -> bytes:
        with self._lock:
            if self._mm is None:
                raise RuntimeError("mmap not initialized")
            head = struct.unpack("<I", self._mm[32:36])[0]
            tail = struct.unpack("<I", self._mm[36:40])[0]
            closed = struct.unpack("<I", self._mm[40:44])[0]
            if closed:
                raise RuntimeError("ring closed")
            if head == tail:
                raise RuntimeError("ring empty")
            offset = HEADER_SIZE + (tail % self.capacity)
            length = struct.unpack("<I", self._mm[offset:offset + 4])[0]
            if length > self.capacity - 4:
                raise RuntimeError("invalid message length")
            if offset + 4 + length > HEADER_SIZE + self.capacity:
                first = HEADER_SIZE + self.capacity - offset - 4
                first_part = self._mm[offset + 4:HEADER_SIZE + self.capacity]
                second_part = self._mm[HEADER_SIZE:HEADER_SIZE + length - first]
                payload = first_part + second_part
            else:
                payload = bytes(self._mm[offset + 4:offset + 4 + length])
            self._mm[36:40] = struct.pack("<I", (tail + 4 + length) % self.capacity)
            return payload

    def close(self) -> None:
        with self._lock:
            if self._mm is None:
                return
            self._mm[40:44] = struct.pack("<I", 1)
            self._mm.close()
            self._mm = None

File: x-runtime/templates/typescript/transport.ts
import * as fs from "fs";
import * as path from "path";

const RING_MAGIC = Buffer.from("HMRING00");
const RING_VERSION = 1;
const HEADER_SIZE = 128;
const MIN_CAPACITY = 4096;

export interface TransportCapabilities {
  supportsZeroCopy: boolean;
  maxMessageBytes: number;
  ringPages: number;
}

export class ZeroCopyRing {
  private fd: number;
  private mmap: Buffer;
  public readonly capacity: number;
  private closed = false;

  constructor(private readonly filePath: string, capacity = MIN_CAPACITY) {
    this.capacity = Math.max(MIN_CAPACITY, capacity);
    this.fd = fs.openSync(filePath, fs.existsSync(filePath) ? "r+" : "w+");
    try {
      fs.ftruncateSync(this.fd, HEADER_SIZE + this.capacity);
      this.mmap = fs.readFileSync(filePath);
      this.writeHeader(HEADER_SIZE + this.capacity);
    } catch (error) {
      fs.closeSync(this.fd);
      throw error;
    }
  }

  private writeHeader(ringSize: number) {
    const header = Buffer.alloc(HEADER_SIZE);
    header.write(RING_MAGIC.toString("binary"), 0, "binary");
    header.writeUInt32LE(RING_VERSION, 8);
    header.writeUInt32LE(0, 12);
    header.writeBigUInt64LE(BigInt(ringSize), 16);
    header.writeUInt32LE(this.capacity, 24);
    header.writeUInt32LE(0, 28); // head
    header.writeUInt32LE(0, 32); // tail
    header.writeUInt32LE(0, 36); // closed
    header.writeBigUInt64LE(0n, 40); // error
    header.writeUInt32LE(1, 48); // supports_zero_copy
    header.writeUInt32LE(this.capacity, 52); // max_message_bytes
    header.writeUInt32LE(1, 56); // ring_pages
    fs.writeSync(this.fd, header, 0, header.length, 0);
    this.mmap = fs.readFileSync(this.filePath);
  }

  static open(filePath: string): ZeroCopyRing {
    const fd = fs.openSync(filePath, "a+");
    const size = fs.fstatSync(fd).size;
    if (size < HEADER_SIZE + MIN_CAPACITY) {
      fs.closeSync(fd);
      throw new Error("ring too small");
    }
    const mmap = fs.readFileSync(filePath);
    const magic = mmap.subarray(0, 8).toString("binary");
    const version = mmap.readUInt32LE(8);
    if (magic !== RING_MAGIC.toString("binary") || version !== RING_VERSION) {
      fs.closeSync(fd);
      throw new Error("invalid ring magic/version");
    }
    const ring = new ZeroCopyRing(filePath, 0);
    ring.fd = fd;
    ring.mmap = mmap;
    ring.capacity = mmap.readUInt32LE(24);
    return ring;
  }

  send(payload: Buffer): void {
    if (payload.length > this.capacity - 4) {
      throw new Error("payload too large");
    }
    const head = this.mmap.readUInt32LE(28);
    const tail = this.mmap.readUInt32LE(32);
    if (this.mmap.readUInt32LE(36) !== 0) {
      throw new Error("ring closed");
    }
    const needed = 4 + payload.length;
    const free = (this.capacity - 1 - ((head - tail) % this.capacity)) % this.capacity;
    if (free < needed) {
      throw new Error("ring full");
    }
    const offset = HEADER_SIZE + (head % this.capacity);
    const end = offset + needed;
    if (end <= HEADER_SIZE + this.capacity) {
      const header = Buffer.alloc(4);
      header.writeUInt32LE(payload.length, 0);
      fs.writeSync(this.fd, header, 0, 4, offset);
      fs.writeSync(this.fd, payload, 0, payload.length, offset + 4);
    } else {
      const first = HEADER_SIZE + this.capacity - offset;
      const header = Buffer.alloc(4);
      header.writeUInt32LE(payload.length, 0);
      fs.writeSync(this.fd, header, 0, 4, offset);
      fs.writeSync(this.fd, payload.subarray(0, first - 4), 0, first - 4, offset + 4);
      fs.writeSync(this.fd, payload.subarray(first - 4), 0, payload.length - (first - 4), HEADER_SIZE);
    }
    const newHead = head + needed;
    const headBuf = Buffer.alloc(4);
    headBuf.writeUInt32LE(newHead, 0);
    fs.writeSync(this.fd, headBuf, 0, 4, 28);
    this.mmap = fs.readFileSync(this.filePath);
  }

  receive(): Buffer {
    const head = this.mmap.readUInt32LE(28);
    const tail = this.mmap.readUInt32LE(32);
    if (this.mmap.readUInt32LE(36) !== 0) {
      throw new Error("ring closed");
    }
    if (head === tail) {
      throw new Error("ring empty");
    }
    const offset = HEADER_SIZE + (tail % this.capacity);
    const length = this.mmap.readUInt32LE(offset);
    if (length > this.capacity - 4) {
      throw new Error("invalid message length");
    }
    const end = offset + 4 + length;
    let payload: Buffer;
    if (end > HEADER_SIZE + this.capacity) {
      const first = HEADER_SIZE + this.capacity - offset - 4;
      payload = Buffer.concat([
        this.mmap.subarray(offset + 4, HEADER_SIZE + this.capacity),
        this.mmap.subarray(HEADER_SIZE, HEADER_SIZE + length - first),
      ]);
    } else {
      payload = this.mmap.subarray(offset + 4, offset + 4 + length);
    }
    const newTail = (tail + 4 + length) % this.capacity;
    const tailBuf = Buffer.alloc(4);
    tailBuf.writeUInt32LE(newTail, 0);
    fs.writeSync(this.fd, tailBuf, 0, 4, 32);
    this.mmap = fs.readFileSync(this.filePath);
    return Buffer.from(payload);
  }

  close(): void {
    if (this.closed) return;
    this.closed = true;
    const closedBuf = Buffer.alloc(4);
    closedBuf.writeUInt32LE(1, 0);
    try {
      fs.writeSync(this.fd, closedBuf, 0, 4, 36);
    } catch {
      // ignore
    }
    try {
      fs.closeSync(this.fd);
    } catch {
      // ignore
    }
  }
}
